stratify/_bounded_vinterp.py

```diff
@@ -0,0 +1,133 @@
+    src_shape.pop(axis_relative)
+    tgt_shape = list(z_target.shape)
+    tgt_shape.pop(axis_relative)
+
+    if src_shape != tgt_shape:
+        src_shape = list(z_src.shape)
+        src_shape[axis_relative] = '-'
+        tgt_shape = list(z_target.shape)
+        src_shape[axis_relative] = '-'
+        msg = ('Expecting the shape of the source and target levels except '
+               'the axis of interpolation to be identical.  {} != {}')
+        raise ValueError(msg.format(tuple(src_shape), tuple(tgt_shape)))
+
+    dat_shape = list(fz_src.shape)
+    dat_shape = dat_shape[-(z_src.ndim-1):]
+    src_shape = list(z_src.shape[:-1])
+    if dat_shape != src_shape:
+        dat_shape = list(fz_src.shape)
+        dat_shape[:-(z_src.ndim-1)] = '-'
+        msg = ('The provided data is not of compatible shape with the '
+               'provided source bounds. {} != {}')
+        raise ValueError(msg.format(tuple(dat_shape), tuple(src_shape)))
+
+    if z_src.shape[-1] != 2:
+        msg = 'Unexpected source and target bounds shape. shape[-1] != 2'
+        raise ValueError(msg)
+
+    # Define our source in a consistent way.
+    # [broadcasting_dims, axis_interpolation, z_varying]
+
+    # src_data
+    bdims = list(range(fz_src.ndim - (z_src.ndim-1)))
+    data_vdims = [ind for ind in range(fz_src.ndim) if ind not in
+                  (bdims + [axis])]
+    data_transpose = bdims + [axis] + data_vdims
+    fz_src_reshaped = np.transpose(fz_src, data_transpose)
+    fz_src_orig = list(fz_src_reshaped.shape)
+    shape = (
+        int(np.product(fz_src_reshaped.shape[:len(bdims)])),
+        fz_src_reshaped.shape[len(bdims)],
+        int(np.product(fz_src_reshaped.shape[len(bdims)+1:])))
+    fz_src_reshaped = fz_src_reshaped.reshape(shape)
+
+    # Define our src and target bounds in a consistent way.
+    # [axis_interpolation, z_varying, 2]
+    vdims = list(set(range(z_src.ndim)) - set([axis_relative]))
+    z_src_reshaped = np.transpose(z_src, [axis_relative] + vdims)
+    z_target_reshaped = np.transpose(z_target, [axis_relative] + vdims)
+
+    shape = int(np.product(z_src_reshaped.shape[1:-1]))
+    z_src_reshaped = z_src_reshaped.reshape([z_src_reshaped.shape[0], shape,
+                                             z_src_reshaped.shape[-1]])
+    shape = int(np.product(z_target_reshaped.shape[1:-1]))
+    z_target_reshaped = z_target_reshaped.reshape(
+        [z_target_reshaped.shape[0], shape, z_target_reshaped.shape[-1]])
+
+    result = conservative_interpolation(
+        z_src_reshaped, z_target_reshaped, fz_src_reshaped)
+
+    # Turn the result into a shape consistent with the source.
+    # First reshape, then reverse transpose.
+    shape = fz_src_orig
+    shape[len(bdims)] = z_target.shape[axis_relative]
+    result = result.reshape(shape)
+    invert_transpose = [data_transpose.index(ind) for ind in
+                        list(range(result.ndim))]
+    result = result.transpose(invert_transpose)
+    return result
```

Comment:
Comments:

 * Looks like this could be a class with behaviour (```conservative_interpolation```).
 * Very similar code for the non-bounded case. Would be interesting to pull this all out into common functions.

Reply:
Agreed.

I saw these two points as being connected.  I think a refactor to support caching would entail pulling both existing and conservative capability into classes.  I didn't want to pre-define a structure before considering this refactor in that wider context of caching (it was easier to use a proxy to the existing API for now).  With regards to common functions, as far as I can tell. the existing stratify usage is rather hardcoded to a points based approach right now.
In the refactor to support caching, I hope to have a similar mindset to what I have done here, in that we pull out the code which needn't be cython code into a python module (inc. enduser API), allowing greater flexibility and functionality to be more readily shared between points-based and bounds-based approaches.  Are you happy with this mindset?

Cheers